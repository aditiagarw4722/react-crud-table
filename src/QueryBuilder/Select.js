// @flow

import React, { Component } from 'react';

type Option = {
  key: string | number,
  value: string | number,
  text: string,
};
type Props = {
  value: ?string | ?number,
  options: Array<Option>,
  placeholder: string,
  onChange: Function,
};

const Select = (props: Props): React$Element<'select'> => {
  const { placeholder, options, value, onChange } = props;
  return (
    <select onChange={onChange}>
      <option value="" selected={!value}>
        {placeholder}
      </option>
      {options.map((option) => (
        <option key={option.key} value={option.value}>
          {option.text}
        </option>
      ))}
    </select>
  );
};

export default Select;
