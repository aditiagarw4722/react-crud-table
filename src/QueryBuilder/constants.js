export const CONDITIONS_LABEL = {
  EQUALS_TO: 'Equals to',
  BEGINS_WITH: 'Begins with',
  ENDS_WITH: 'Ends with',
  CONTAINS: 'Contains',
  LESS_THAN: 'Less than',
  GREATER_THAN: 'Greater than',
  LESS_OR_EQUALS_THAN: 'Less or equals than',
  GREATER_OR_EQUALS_THAN: 'Greater or equals than',
  IS: 'Is',
  IS_NOT: 'Is not',
  IS_NOT_EQUALS_TO: 'IS not equals to',
  DOES_NOT_BEGIN_WITH: ' Does not begin with',
  DOES_NOT_END_WITH: ' Does not end with',
  DOES_NOT_CONTAIN: ' Does not contain',
  IS_NOT_LESS_THAN: 'Is not less than',
  IS_NOT_GREATER_THAN: 'IS not greater than',
  IS_NOT_LESS_OR_EQUALS_THAN: 'Is not less or equals than',
  IS_NOT_GREATER_OR_EQUALS_THAN: 'IS not greater or equals than',
};

export const CONDITIONS = {
  EQUALS_TO: 'EQUALS_TO',
  BEGINS_WITH: 'BEGINS_WITH',
  ENDS_WITH: 'ENDS_WITH',
  CONTAINS: 'CONTAINS',
  LESS_THAN: 'LESS_THAN',
  GREATER_THAN: 'GREATER_THAN',
  LESS_OR_EQUALS_THAN: 'LESS_OR_EQUALS_THAN',
  GREATER_OR_EQUALS_THAN: 'GREATER_OR_EQUALS_THAN',
  IS: 'IS',
  IS_NOT: 'IS_NOT',
  IS_NOT_EQUALS_TO: 'IS_NOT_EQUALS_TO',
  DOES_NOT_BEGIN_WITH: 'DOES_NOT_BEGIN_WITH',
  DOES_NOT_END_WITH: 'DOES_NOT_END_WITH',
  DOES_NOT_CONTAIN: 'DOES_NOT_CONTAIN',
  IS_NOT_LESS_THAN: 'IS_NOT_LESS_THAN',
  IS_NOT_GREATER_THAN: 'IS_NOT_GREATER_THAN',
  IS_NOT_LESS_OR_EQUALS_THAN: 'IS_NOT_LESS_OR_EQUALS_THAN',
  IS_NOT_GREATER_OR_EQUALS_THAN: 'IS_NOT_GREATER_OR_EQUALS_THAN',
};

export const DEFAULT_STATE = {
  field: '',
  value: '',
  type: '',
  condition: '',
  collection: '',
};

export default {
  CONDITIONS_LABEL,
  CONDITIONS,
  DEFAULT_STATE,
};
