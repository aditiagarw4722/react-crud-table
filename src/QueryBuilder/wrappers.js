import bcc from 'bem-react-component-creator';

const { block, element } = bcc('crud-table-query-builder');

export const Container = block('div');

Container.displayName = 'Container';

export const RuleBuilder = element('div', 'search-rule-builder');

RuleBuilder.displayName = 'RuleBuilder';

export default { Container, RuleBuilder };
