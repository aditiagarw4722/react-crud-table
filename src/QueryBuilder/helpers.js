import React from 'react';
import DateTime from 'react-datetime';
import Label from '../Label';
import { CONDITIONS_LABEL, CONDITIONS } from './constants';

let styles;

const Input = (props) => <input {...props} />;

export const isRuleComplete = (rule) => {
  const { condition, value, field } = rule;
  return condition !== '' && value !== '' && field !== '';
};

export const isBoolean = (type) => type === 'boolean';

export const mapFieldsToOptions = (fields) =>
  fields.map((x) => ({
    text: x.label,
    value: x.value,
    key: x.value,
  }));

export const defaultRuleRender = (rule) => {
  if (isBoolean(rule.type)) {
    return (
      <span>
        <Label style={styles.label}>{rule.condition}</Label>
        &nbsp;
        <Label style={styles.label}>{rule.label}</Label>
      </span>
    );
  }
  return (
    <span>
      <Label style={styles.label}>{rule.label}</Label>
      &nbsp;
      <Label style={styles.label}>{CONDITIONS_LABEL[rule.condition]}</Label>
      &nbsp;
      <Label style={styles.label}>{rule.value}</Label>
    </span>
  );
};

export const conditionsForType = (type) => {
  let result;
  switch (type) {
    case 'number':
    case 'date':
      result = [
        CONDITIONS.EQUALS_TO,
        CONDITIONS.LESS_THAN,
        CONDITIONS.GREATER_THAN,
        CONDITIONS.LESS_OR_EQUALS_THAN,
        CONDITIONS.GREATER_OR_EQUALS_THAN,
        CONDITIONS.IS_NOT_EQUALS_TO,
        CONDITIONS.IS_NOT_LESS_THAN,
        CONDITIONS.IS_NOT_GREATER_THAN,
        CONDITIONS.IS_NOT_LESS_OR_EQUALS_THAN,
        CONDITIONS.IS_NOT_GREATER_OR_EQUALS_THAN,
      ];
      break;
    case 'boolean':
      result = [CONDITIONS.IS, CONDITIONS.IS_NOT];
      break;
    default:
      result = [
        CONDITIONS.CONTAINS,
        CONDITIONS.EQUALS_TO,
        CONDITIONS.BEGINS_WITH,
        CONDITIONS.ENDS_WITH,
        CONDITIONS.IS_NOT_EQUALS_TO,
        CONDITIONS.DOES_NOT_BEGIN_WITH,
        CONDITIONS.DOES_NOT_END_WITH,
        CONDITIONS.DOES_NOT_CONTAIN,
      ];
      break;
  }

  return result.map((r) => ({
    value: r,
    text: CONDITIONS_LABEL[r],
    key: r,
  }));
};

export const getDefaultConditionForType = (type) => {
  switch (type) {
    case 'number':
      return CONDITIONS.EQUALS_TO;
    case 'date':
      return CONDITIONS.GREATER_OR_EQUALS_THAN;
    default:
      return CONDITIONS.CONTAINS;
  }
};

export const inputForType = (type, props) => {
  switch (type) {
    case 'number':
      return <Input type="number" {...props} />;
    case 'boolean':
      return (
        <Input
          {...props}
          type="checkbox"
          onClick={(evt) => {
            props.onChange({
              ...evt,
              target: {
                ...evt.target,
                value: evt.target.checked,
              },
            });
          }}
          onChange={() => {}}
        />
      );
    case 'date':
      return (
        <DateTime
          {...props}
          className="ui input"
          dateFormat="YYYY-MM-DD"
          timeFormat="hh:mm A"
          onChange={(data) => {
            props.onChange({
              target: {
                value: data.format('YYYY-MM-DD hh:mm A'),
              },
            });
          }}
        />
      );
    default:
      return <Input type="text" {...props} />;
  }
};

styles = {
  label: {
    marginBottom: '5px',
  },
};

export default {
  defaultRuleRender,
  conditionsForType,
  getDefaultConditionForType,
  inputForType,
  mapFieldsToOptions,
  isBoolean,
  isRuleComplete,
};
