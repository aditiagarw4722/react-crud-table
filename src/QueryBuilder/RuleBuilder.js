// @flow

import React from 'react';

import Button from '../Button';
import Select from '../Select';
import { queryValue } from '../CRUDTable/helpers';
import {
  conditionsForType,
  getDefaultConditionForType,
  inputForType,
  mapFieldsToOptions,
  isBoolean,
  isRuleComplete,
} from './helpers';
import { CONDITIONS, DEFAULT_STATE } from './constants';
import { RuleBuilder as Container } from './wrappers';

type Field = {
  value: any,
};
type Props = {
  fields: Array<Field>,
  conditionsSelectPlaceholder: string,
  fieldsSelectPlaceholder: string,
  onSave: Function,
};

type State = {
  field: any,
  type: string,
  value: any,
  label: string,
  collection: string,
  condition: string,
};

type Data = {
  value: any,
};

class RuleBuilder extends React.Component<Props, State> {
  constructor(props: Props) {
    super(props);
    this.state = DEFAULT_STATE;
    // $FlowFixMe
    this.handleFieldSelectChange = this.handleFieldSelectChange.bind(this);
    // $FlowFixMe
    this.save = this.save.bind(this);
  }

  handleFieldSelectChange(evt: SyntheticEvent<HTMLSelectElement>) {
    const { value } = evt.currentTarget;
    const update: State = {
      ...this.state,
      field: value,
      type: this.getType(value),
      label: this.getLabel(value),
      collection: this.getCollection(value),
    };

    update.condition = this.getDefaultCondition(value);

    if (isBoolean(update.type)) {
      update.condition = CONDITIONS.IS;
      update.value = false;
    }

    this.setState(update);
  }

  getLabel(field: any): string {
    return queryValue(this.find(field), 'label', '');
  }

  getDefaultCondition(field: any): string {
    const defaultconditionForType = getDefaultConditionForType(
      this.getType(field)
    );

    return queryValue(
      this.find(field),
      'defaultCondition',
      defaultconditionForType
    );
  }

  getCollection(field: any): string {
    return queryValue(this.find(field), 'collection', '');
  }

  getType(field: any): string {
    return queryValue(this.find(field), 'type', '');
  }

  find(field: any): ?Field {
    const { fields } = this.props;
    return fields.find((f) => f.value === field);
  }

  save() {
    const { onSave } = this.props;
    if (isRuleComplete(this.state)) {
      onSave(this.state);
      this.setState(DEFAULT_STATE);
    }
  }

  render(): React$Element<any> {
    const {
      fields,
      conditionsSelectPlaceholder,
      fieldsSelectPlaceholder,
    } = this.props;
    const { field, value, condition } = this.state;
    const type = this.getType(field);
    const input = inputForType(type, {
      value,
      onChange: (evt) => {
        this.setState({ value: evt.currentTarget.value });
      },
    });

    return (
      <Container>
        <Select
          placeholder={fieldsSelectPlaceholder}
          options={mapFieldsToOptions(fields)}
          value={field}
          onChange={this.handleFieldSelectChange}
        />
        &nbsp;&nbsp;
        {!isBoolean(type) && (
          <span>
            <Select
              placeholder={conditionsSelectPlaceholder}
              options={conditionsForType(type)}
              value={condition}
              onChange={(evt, data) => {
                this.setState({ condition: evt.currentTarget.value });
              }}
            />
            &nbsp;&nbsp;
          </span>
        )}
        {input}
        &nbsp;&nbsp;
        <Button modifiers="positive,add" onClick={this.save}>
          +
        </Button>
      </Container>
    );
  }
}

export default RuleBuilder;
