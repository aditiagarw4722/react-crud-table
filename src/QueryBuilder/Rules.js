import React from 'react';
import PropTypes from 'prop-types';
import { defaultRuleRender } from './helpers';
import Button from '../Button';
import { NO_OP } from '../helpers';

const Rules = ({ queryRules, onRuleRemoved, renderRule }) => (
  <div>
    {queryRules.map((rule) => (
      <div key={`${rule.field}:${rule.condition}`}>
        {renderRule(rule)}
        {'  '}
        <Button
          onClick={() => {
            onRuleRemoved(rule);
          }}
          modifiers="negative,circular"
        >
          X
        </Button>
      </div>
    ))}
    <br />
  </div>
);

Rules.propTypes = {
  queryRules: PropTypes.instanceOf(Array),
  onRuleRemoved: PropTypes.func,
  renderRule: PropTypes.func,
};

Rules.defaultProps = {
  renderRule: defaultRuleRender,
  queryRules: [],
  onRuleRemoved: NO_OP,
};

export default Rules;
