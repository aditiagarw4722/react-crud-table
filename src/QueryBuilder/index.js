// @flow

import React from 'react';
import RuleBuilder from './RuleBuilder';
import Rules from './Rules';
import { Container } from './wrappers';

type Props = {
  queryRules: Array<Object>,
  fields: Array<Object>,
  renderRule?: Function,
  onRuleAdded: Function,
  onRuleRemoved: Function,
};

const QueryBuilder = (props: Props): React$Element<any> => {
  const { queryRules, fields, renderRule, onRuleAdded, onRuleRemoved } = props;
  return (
    <Container>
      <RuleBuilder
        fields={fields}
        onSave={onRuleAdded}
        fieldsSelectPlaceholder="Select field"
        conditionsSelectPlaceholder="Select condition"
      />
      <Rules
        queryRules={queryRules}
        onRuleRemoved={onRuleRemoved}
        renderRule={renderRule}
      />
    </Container>
  );
};

export default QueryBuilder;
