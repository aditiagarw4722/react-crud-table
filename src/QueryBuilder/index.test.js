import React from 'react';
import ShallowRenderer from 'react-test-renderer/shallow';
import { create } from 'react-test-renderer';

import RuleBuilder from './RuleBuilder';
import Rules from './Rules';
import QueryBuilder from './';

const renderer = new ShallowRenderer();

describe('QueryBuilder', () => {
  const rule = {
    field: 'name',
    condition: 'EQUALS_TO',
  };

  const props = {
    fields: [
      {
        name: 'id',
        value: 1,
      },
    ],
    queryRules: [rule],
    renderRule: jest.fn(),
    onRuleAdded: jest.fn(),
    onRuleRemoved: jest.fn(),
    onChange: jest.fn(),
  };

  it('renders as expected with the given props', () => {
    const result = renderer.render(<QueryBuilder {...props} />);
    expect(result).toMatchSnapshot();
  });

  it('notifies when a rule is saved', () => {
    const result = create(<QueryBuilder {...props} />);

    result.root.findByType(RuleBuilder).props.onSave(rule);

    expect(props.onRuleAdded).toHaveBeenCalledTimes(1);
    expect(props.onRuleAdded).toHaveBeenCalledWith(rule);
  });

  it('notifies when a rule is removed', () => {
    const result = create(<QueryBuilder {...props} />);

    result.root.findByType(Rules).props.onRuleRemoved(rule);

    expect(props.onRuleRemoved).toHaveBeenCalledTimes(1);
    expect(props.onRuleRemoved).toHaveBeenCalledWith(rule);
  });
});
