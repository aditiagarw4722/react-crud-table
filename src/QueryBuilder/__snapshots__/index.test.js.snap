// Jest Snapshot v1, https://goo.gl/fbAQLP

exports[`QueryBuilder renders as expected with the given props 1`] = `
<Container>
  <RuleBuilder
    conditionsSelectPlaceholder="Select condition"
    fields={
      Array [
        Object {
          "name": "id",
          "value": 1,
        },
      ]
    }
    fieldsSelectPlaceholder="Select field"
    onSave={[MockFunction]}
  />
  <Rules
    onRuleRemoved={[MockFunction]}
    queryRules={
      Array [
        Object {
          "condition": "EQUALS_TO",
          "field": "name",
        },
      ]
    }
    renderRule={[MockFunction]}
  />
</Container>
`;
